storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,671 @@
+    ///
+    /// The `concurrency` parameter controls how many blobs are replayed concurrently. This can
+    /// dramatically speed up the replay process if the underlying storage supports concurrent reads
+    /// across different blobs.
+    pub async fn replay(
+        &mut self,
+        concurrency: usize,
+    ) -> Result<impl Stream<Item = Result<(u64, [u8; N]), Error>> + '_, Error> {
+        // Collect all blobs to replay
+        let mut blobs = Vec::with_capacity(self.blobs.len());
+        for (index, blob) in self.blobs.iter().enumerate() {
+            let blob_len = {
+                if index == self.blobs.len() - 1 {
+                    blob.len().await?
+                } else {
+                    self.cfg.items_per_blob * (size_of::<u32>() + N) as u64
+                }
+            };
+            blobs.push((index, blob, blob_len));
+        }
+
+        // Replay all blobs concurrently and stream items as they are read (to avoid occupying too
+        // much memory with buffered data)
+        let items_per_blob = self.cfg.items_per_blob;
+        let pruned_items = self.oldest_unpruned_blob as u64 * items_per_blob;
+        Ok(stream::iter(blobs)
+            .map(move |(index, blob, blob_len)| async move {
+                stream::unfold(
+                    (index, blob, 0u64),
+                    move |(index, blob, offset)| async move {
+                        // Check if we are at the end of the blob
+                        if offset == blob_len {
+                            return None;
+                        }
+                        // Get next item
+                        let mut buf = vec![0u8; N + size_of::<u32>()];
+                        let item = blob.read_at(&mut buf, offset).await.map_err(Error::Runtime);
+                        let next_offset = offset + (size_of::<u32>() + N) as u64;
+                        match item {
+                            Ok(_) => match Self::verify_integrity(&buf) {
+                                Ok(item) => Some((
+                                    Ok((
+                                        pruned_items
+                                            + items_per_blob * index as u64
+                                            + offset / (size_of::<u32>() + N) as u64,
+                                        item,
+                                    )),
+                                    (index, blob, next_offset),
+                                )),
+                                Err(err) => Some((Err(err), (index, blob, next_offset))),
+                            },
+                            Err(err) => Some((Err(err), (index, blob, blob_len))),
+                        }
+                    },
+                )
+            })
+            .buffer_unordered(concurrency)
+            .flatten())
+    }
+
+    fn most_recent_blob_number(&self) -> u64 {
+        self.blobs.len() as u64 + self.oldest_unpruned_blob as u64 - 1
```

Comment:
Might be worth commenting why this can never be negative (would panic)?

Reply:
done